packages/template-retail-react-app/app/commerce-hooks/context/shopper-basket-context.js

```diff
@@ -0,0 +1,126 @@
+import basketsReducer from '../reducers/baskets-reducer'
+
+const basketsInitialState = {
+    isLoading: false,
+    error: false,
+    data: [],
+    total: 0
+}
+const BasketsContext = React.createContext()
+const BasketsProvider = ({children}) => {
+    // TODO: we should separate api into different apis context instead of having a combined api object here
+    // for example: ShopperBasketAPI, ShopperCustomerAPI
+    // expect to use const api = shopperBasketAPI()
+    const api = useCommerceAPI()
+
+    // slas hooks
+    const shopper = useCustomer()
+    // Create or restore the user session upon mounting
+    // this flow should stay in shopper hook
+    useEffect(() => {
+        shopper.login()
+    }, [])
+
+    const [baskets, dispatch] = React.useReducer(basketsReducer, basketsInitialState)
+
+    const value = React.useMemo(() => ({baskets, dispatch}), [baskets])
+
+    const createBasket = async () => {
+        dispatch({type: 'loading'})
+        try {
+            const res = await api.shopperBaskets.createBasket({})
+            console.log('seting a basket====================')
+            dispatch({type: 'set_basket', payload: res})
+            return res
+        } catch (err) {
+            dispatch({type: 'error'})
+            throw new Error(err)
+        }
+    }
+
+    const getBaskets = async () => {
+        dispatch({type: 'loading'})
+        try {
+            console.log('getting baskets---------------------------')
+            console.log('api.shopperBaskets', api.shopperBaskets)
+            const res = await api.shopperCustomers.getCustomerBaskets({
+                parameters: {customerId: shopper?.customerId}
+            })
+            console.log('res', res)
+            if (!res.total) {
+                console.log('creating a basket ----------------------------')
+                await createBasket()
+            }
+
+            console.log('setting baskets-------------------------')
+            dispatch({type: 'set_baskets', payload: res})
+        } catch (e) {
+            console.log('basket error --------------------')
+            console.log('e', e)
+        }
+    }
+
+    const addItemToCart = async (item, basketId) => {
+        console.log('add to cart loading ----')
+        dispatch({type: 'loading'})
+        try {
+            console.log('adding-item to cart')
+            const res = await api.shopperBaskets.addItemToBasket({
+                body: item,
+                parameters: {basketId}
+            })
+            console.log('res', res)
+            dispatch({type: 'add_to_cart', payload: res})
+        } catch (err) {
+            dispatch({type: 'error'})
+            throw new Error(err)
+        }
+    }
+
+    // retrieve the baskets information on first render
+    // if there is no basket, create one => can we delay this until someone adds an item to cart
+    React.useEffect(() => {
+        console.log('Initializing baskets =========================')
```

Comment:
Thought: we probably can delay this action further into an item is added to cart to avoid delaying any rendering time for pages